packages/matter.js/src/protocol/interaction/AttributeDataEncoder.ts

```diff
@@ -0,0 +1,249 @@
+    eventData?: EventDataPayload;
+};
+
+/** Type for TlvEventData where the real data are represented with the schema and the JS value. */
+type EventDataPayload = Omit<TypeFromSchema<typeof TlvEventData>, "data"> & {
+    schema: TlvSchema<any>;
+    payload: any;
+};
+
+/** Type for TlvDataReport where the real data are represented with the schema and the JS value. */
+export type DataReportPayload = Omit<TypeFromSchema<typeof TlvDataReport>, "attributeReports" | "eventReports"> & {
+    attributeReportsPayload?: AttributeReportPayload[];
+    eventReportsPayload?: EventReportPayload[];
+};
+
+/** Encodes an AttributeReportPayload into a TlvStream (used for TlvAny type). */
+export function encodeAttributePayload(attributePayload: AttributeReportPayload): TlvStream {
+    const { attributeData, attributeStatus } = attributePayload;
+    if (attributeData === undefined) {
+        return TlvAttributeReport.encodeTlv({ attributeStatus });
+    }
+
+    const { path, schema, payload, dataVersion } = attributeData;
+    return TlvAttributeReport.encodeTlv({ attributeData: { path, data: schema.encodeTlv(payload), dataVersion } });
+}
+
+/** Encodes an EventReportPayload into a TlvStream (used for TlvAny type). */
+export function encodeEventPayload(eventPayload: EventReportPayload): TlvStream {
+    const { eventData, eventStatus } = eventPayload;
+    if (eventData === undefined) {
+        return TlvEventReport.encodeTlv({ eventStatus });
+    }
+
+    const {
+        path,
+        schema,
+        payload,
+        eventNumber,
+        deltaEpochTimestamp,
+        epochTimestamp,
+        deltaSystemTimestamp,
+        systemTimestamp,
+        priority,
+    } = eventData;
+    return TlvEventReport.encodeTlv({
+        eventData: {
+            path,
+            data: schema.encodeTlv(payload),
+            priority,
+            systemTimestamp,
+            deltaSystemTimestamp,
+            deltaEpochTimestamp,
+            epochTimestamp,
+            eventNumber,
+        },
+    });
+}
+
+/** Return if an AttributeReportPayload can be chunked or not. */
+export function canAttributePayloadBeChunked(attributePayload: AttributeReportPayload): boolean {
+    const { attributeData } = attributePayload;
+    if (attributeData === undefined) {
+        return false;
+    }
+    const {
+        schema,
+        payload,
+        path: { listIndex },
+    } = attributeData;
+    return schema instanceof ArraySchema && Array.isArray(payload) && payload.length > 0 && listIndex === undefined;
+}
+
+/** Chunk an AttributeReportPayload into multiple AttributeReportPayloads. */
+export function chunkAttributePayload(attributePayload: AttributeReportPayload): AttributeReportPayload[] {
+    const { attributeData } = attributePayload;
+    if (attributeData === undefined) {
+        // Should never happen, but we can make sure it do not break
+        return [attributePayload];
```

Comment:
I'm wondering if it wouldn't be better to break with an error message.  Probably more philosophical than anything else

Reply:
You are right. I will change. Then we see such cases.